File: core/java/src/main/java/hara/lang/protocol/IPeekFirst.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.ipeekfirst", name = "IPeekFirst")
public interface IPeekFirst<E> {
  @HaraMethod(value = "peek-first", arity = 1)
  E peekFirst();
}

File: core/rust/src/lang/protocol/ipeekfirst.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.ipeekfirst", name = "IPeekFirst")]
pub trait IPeekFirst<E> {
    #[hara_method(value = "peek-first", arity = 1)]
    fn peek_first(&self) -> Option<E>;
}
